tests/pfcwd/test_pfcwd_function.py

```diff
@@ -0,0 +1,518 @@
+        if type(self.pfc_wd_test_port_id) == list:
+            dst_port = "".join(str(self.pfc_wd_test_port_id)).replace(',', '')
+        else:
+            dst_port = "[ " + str(self.pfc_wd_test_port_id) + " ]"
+        ptf_params = {'router_mac': self.eth0_mac,
+                      'queue_index': self.pfc_queue_index,
+                      'pkt_count': self.pfc_wd_test_pkt_count,
+                      'port_src': self.pfc_wd_rx_port_id[0],
+                      'port_dst': dst_port,
+                      'ip_dst': self.pfc_wd_test_neighbor_addr,
+                      'port_type': self.port_type,
+                      'wd_action': 'dontcare'}
+        log_format = datetime.datetime.now().strftime("%Y-%m-%d-%H:%M:%S")
+        log_file = "/tmp/pfc_wd.PfcWdTest.{}.log".format(log_format)
+        ptf_runner(self.ptf, "ptftests", "pfc_wd.PfcWdTest", "ptftests", params=ptf_params,
+                   log_file=log_file)
+
+    def verify_wd_func(self, action):
+        """
+        PTF traffic send and verify
+
+        Args:
+            action(string) : PTF traffic test action
+        """
+        logger.info("--- Verify PFCwd function for action {} ---".format(action))
+        self.verify_tx_egress(action)
+        self.verify_rx_ingress(action)
+        self.verify_other_queue()
+        self.verify_other_pg()
+
+
+class TestPfcwdFunc(SetupPfcwdFunc):
+    """ Test PFC function and supporting methods """
+    def storm_detect_path(self, dut, port, action):
+        """
+        Storm detection action and associated verifications
+
+        Args:
+            dut(AnsibleHost) : DUT instance
+            port(string) : DUT port
+            action(string) : PTF test action
+
+        Returns:
+            loganalyzer(Loganalyzer) : instance
+        """
+        restore_time = self.timers['pfc_wd_restore_time_large']
+        detect_time = self.timers['pfc_wd_detect_time']
+
+        loganalyzer = LogAnalyzer(ansible_host=self.dut,
+                                  marker_prefix="pfc_function_storm_detect_{}_port_{}".format(action, port))
+        marker = loganalyzer.init()
+        ignore_file = os.path.join(TEMPLATES_DIR, "ignore_pfc_wd_messages")
+        reg_exp = loganalyzer.parse_regexp_file(src=ignore_file)
+        loganalyzer.ignore_regex.extend(reg_exp)
+        loganalyzer.expect_regex = []
+        loganalyzer.expect_regex.extend([EXPECT_PFC_WD_DETECT_RE])
+        loganalyzer.match_regex = []
+
+        if action != "dontcare":
+            start_wd_on_ports(dut, port, restore_time, detect_time, action)
+
+        self.storm_hndle.start_storm()
+
+        if action == "dontcare":
+            self.traffic_inst.send_dummy_traffic()
+            start_wd_on_ports(dut, port, restore_time, detect_time, "drop")
+
+        time.sleep(5)
+
+        # storm detect
+        logger.info("Verify if PFC storm is detected on port {}".format(port))
+        loganalyzer.analyze(marker)
```

Comment:
Do we have counters that could verify PFC strom took place? if so, we may have follow up PR and change how verification is done.

Reply:
+1. I think checking the logs is a good practice but if have something to cross-check against that would also be great!

Reply:
there is a counter to check if the queue is currently in storm. But its difficult to get the timing right because the PFCwd will restore it as soon as it detects the storm. I will explore that option and add one if possible in a follow up PR. But will still keep this log check because that's the first thing that we generally check if we suspect that there has been a PFC storm